### PR TITLE
Debugging Guide

### PR DESCRIPTION
**Pull request by jamesjuett**
This PR adds a debugging guide for EECS 280 students.

The overall goal is to provide students with a quick survey of effective debugging techniques as well as concrete examples from real EECS 280 project contexts that showcase why they are useful. These generally line up with an introduction to debugger "features".

Note that this guide does **not** intend to tell students how to configure/run the debugger in a particular IDE. We've already got other tutorials for that.

**Preview**  
[https://preview.sesh.rs/previews/eecs280staff/tutorials/27/debugging_guide.html](https://preview.sesh.rs/previews/eecs280staff/tutorials/27/debugging_guide.html)

TODO:
- [ ] Add a link back to relevant portions of other tutorials for setting up and configuring the debugger in each IDE

### CODE REVIEW

File: docs/debugging_guide.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+<img src="images/debug_feature_console_1.png" width="400px" />
+
+That suggests the loop is effectively infinite, and only stops when the pointer travels far enough to segfault. Looking back at my code (see above), I find that the loop condition `ptr < ptr + size` is incorrect.
+
+You can enter almost any valid C++ expression at the debug console - even function calls! Take a look below - we've called `Matrix_print()` from the debug console and the result shows at the terminal.
+
+<img src="images/debug_feature_console_2.png" width="800px" />
+
+### Debug Logging
```
Comment (awdeorio) [inline docs/debugging_guide.md:121]: This is the lightest-weight strategy.  WDTY about moving it up to the first subsection?

File: docs/debugging_guide.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+TODO: link back to relevant parts of IDE or setup tutorials for debuggers
+
+## Inspect Program State
+
+A fundamental part of debugging is simply getting more information about what exactly your program is doing.
+
+### Local Variables
+
+<div class="primer-spec-callout info icon-info" markdown="1">
+Use the **variables panel** to inspect the values of variables while your debugger is paused.
```
Comment (awdeorio) [inline docs/debugging_guide.md:28]: Do you think it's worth mentioning that you can hover the mouse pointer over a variable in the source code?

Does the variables pane advice apply to the other debuggers?  I know that the hover does.

File: docs/debugging_guide.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+        "description": "Print polymorphic objects according to dynamic type",
+        "text": "set print object on"
+      }
+      ...
+  ]
+```
+{: data-variant="legacy" }
+</div>
+
+In **member functions**, you can also open up the `this` pointer. In our Euchre example, we'd see member variables from the current `Game` class instance:
```
Comment (awdeorio) [inline docs/debugging_guide.md:64]: Might be worth a subheading here, possibly hidden from the ToC.  I can image wanting to link to this specific subsection from somewhere in the P3 (or later) spec.

File: docs/debugging_guide.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+<img src="images/debug_crash_implicit.png" width="600px" />
+
+**Tracing Back to _Your_ Code**  
+To diagnose crashes where the debugger initially shows you library code, you need to find the nearest part of _your_ code that ultimately called the library functions.
+
+Let's consider again the case of a crash in an implicitly-defined function. The debugger first just shows us the top line of the class definition, since there's no code to show for implicitly-defined functions.
+
+<img src="images/debug_crash_implicit_full.png" width="800px" />
+
+In this case, check the call stack. First, observe that the segfault did occur in an implicitly-defined function, the built-in `Card::operator=` assignment operator for copying a `Card`.
```
Comment (awdeorio) [inline docs/debugging_guide.md:217]: Opportunity to link back to the above section!

File: docs/debugging_guide.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+Let's run in a debugger and let it crash there. No need to set any breakpoints.
+
+<img src="images/debug_crash_0.png" width="800px" />
+
+From the above, we can quickly see the line where the segfault occurs. We also observe that `p1->get_name()` crashes because `p1` is `0x0` (a null pointer), as seen in the variables panel. Ah ha - we just mistyped `p1` on this line instead of `p2`.
+
+Sometimes you might also need to explore [the call stack](#the-call-stack) to determine the cause of the crash. 
+
+### Crash in _Library/System_ Code
```
Comment (awdeorio) [inline docs/debugging_guide.md:194]: Love this!

File: docs/debugging_guide.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+<img src="images/debug_breakpoint_0.png" width="700px" />
+
+Now, when I run the program via my debugger, it pauses just **before** running that line. That is, the highlighted line is up next, but has not run yet.
+
+<img src="images/debug_breakpoint_1.png" width="700px" />
+
+Now, I can inspect the `summary` variable and see if my `summarize()` function was returning the right result.
+
+<div class="primer-spec-callout warning" markdown="1">
+If your debugger isn't respecting your breakpoints or the lines where it pauses don't seem to match your source code, double check that you've **saved** and **re-compiled** any source files you were editing. (Some IDEs may auto-save or auto-build when you launch the debugger, but not all.)
```
Comment (awdeorio) [inline docs/debugging_guide.md:246]: Another pitfall here is `-g` and not `-O...`.  Optimization makes weird things happen.  I could image 281 students referring back to this tutorial, which is why I mention it.

File: docs/debugging_guide.md [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+<img src="images/debug_icon_stop.png" style="vertical-align: text-top; height: 1.25em;" /> **Stop**  
+Exit the debugger.
+
+### Strategies
+
+Here's a few high-level strategies for navigating effectively when debugging your code.
+
+#### Step Over and Check
+
+<div class="primer-spec-callout info icon-info" markdown="1">
```
Comment (awdeorio) [inline docs/debugging_guide.md:307]: I like this use of callouts in these sections

**Review by awdeorio — APPROVED** [review]
I really like the taxonomy of strategies.  I can imagine pointing students to specific strategies in office hours.  I left some small comments.  The overall document LGTM.

**Comment by sofiazxy** [comment]
This is super useful doc and addresses many key areas in debugging that students would find helpful. A couple of questions - would this guide be used by students or for staff training? If it's the latter, it might be helpful to incorporate  the P5 buggy code training that was put together recently.

File: docs/debugging_guide.md (hunk 1 above)
Comment (jamesjuett) [reply docs/debugging_guide.md:121]: I think I prefer the ordering as-is, with this last. I've found many students actually experience the debugger first when they're trying to track down a crash, in which case being able to inspect program state through the debugger features itself is a first-line approach.

But also, there are a few wording things in that section that would need to be adjusted (e.g. where we contrast against the other approaches) and unless there's a strong reason to move it, it's easier to leave the ordering :).

File: docs/debugging_guide.md (hunk 2 above)
Comment (jamesjuett) [reply docs/debugging_guide.md:28]: Good call! I think this is worth a mention. Added in de6d5b4

Comment (jamesjuett) [reply docs/debugging_guide.md]: Good call! I think this is worth a mention. Added in de6d5b4

Comment (jamesjuett) [reply docs/debugging_guide.md]: I believe all the variables panel advice applies to the three main IDEs/editors we recommend.

File: docs/debugging_guide.md (hunk 3 above)
Comment (jamesjuett) [reply docs/debugging_guide.md:70]: Added in dd7f165

File: docs/debugging_guide.md (hunk 4 above)
Comment (jamesjuett) [reply docs/debugging_guide.md]: Done. c757b03

File: docs/debugging_guide.md (hunk 6 above)
Comment (jamesjuett) [reply docs/debugging_guide.md]: Good call. Added this and others in db40345.

**Comment by jamesjuett** [comment]
> Note: in #30 I borrowed your idea of the little button descriptions and copied your images. I needed to rename them to avoid a name clash with the other IDEs. You might consider doing the same and we can share the same images. `debug_icon_step_over.png` -> `vscode_icon_step_over.png`

Got it! Updated in 1f476ce

**Comment by jamesjuett** [comment]
> This is super useful doc and addresses many key areas in debugging that students would find helpful. A couple of questions - would this guide be used by students or for staff training? If it's the latter, it might be helpful to incorporate the P5 buggy code training that was put together recently.

Intended to be student-facing but I suspect would be useful for our staff to review as well (including so that they can point students to relevant sections of this guide in office hours!).

**Comment by awdeorio** [comment]
To me this looks ready to go!

**Pull request merged**